k2/csrc/array.h

```diff
@@ -454,17 +529,57 @@ class Array2 {
       with the "no-gaps" semantics if dim0_ <= 1. */
   bool IsContiguous() const { return elem_stride0_ == dim1_; }
 
+
+  // Only if this is a Array2<Any>: convert to a specific type, which must be
+  // identical to the actual type stored as dtype_.
+  // You call this as, e.g.
+  //  Array2<Any> a(...);
+  //  if (a.GetDtype() == kInt32Dtype) {
+  //     Array2<int32_t> &b = a.template Specialize<int32_t>();
+  //     ...
+  //  }
+  // Note: the '.template' above is only necessary if you are inside
+  // a template definition already.  It is an odd syntax.
+  template <typename U>
+  Array2<U> &Specialize() {
+    static_assert(std::is_same<T, Any>::value,
+                  "generic arrays not supported here");
+    K2_CHECK_EQ(dtype_, DtypeOf<U>::dtype);
+    return *reinterpret_cast<Array2<U>*>(this);
+  }
+  template <typename U>
+  const Array2<U> &Specialize() const {
+    static_assert(std::is_same<T, Any>::value,
+                  "generic arrays not supported here");
+    K2_CHECK_EQ(dtype_, DtypeOf<U>::dtype);
+    return *reinterpret_cast<const Array2<U>*>(this);
+  }
+
+  // Return a reference to this viewed as `Any` type (for when we
+  // want a generic array without type informatiopn)
+  // Reverse this with Specialize().
+  Array2<Any> &Generic() {
+    return *reinterpret_cast<Array2<Any>*>(this);
+  }
+  Array2<Any> &Generic() const {
+    return *reinterpret_cast<const Array2<Any>*>(this);
+  }
+
+
   /*  returns a flat version of this, appending the rows; will copy the data if
       it was not contiguous. */
   Array1<T> Flatten() {
     NVTX_RANGE(K2_FUNC);
     if (dim1_ == elem_stride0_) {
-      return Array1<T>(dim0_ * dim1_, region_, byte_offset_);
+      return Array1<T>(dim0_ * dim1_, region_, byte_offset_,
+                       dtype_);
     } else {
       auto region = NewRegion(region_->context, static_cast<size_t>(dim0_) *
-                                                    static_cast<size_t>(dim1_) *
-                                                    ElementSize());
-      Array1<T> array(dim0_ * dim1_, region, 0);
+                              static_cast<size_t>(dim1_) *
+                              ElementSize());
+      Array1<T> array(dim0_ * dim1_, region, 0, dtype_);
+      static_assert(!K2_TYPE_IS_ANY(T),
```

Comment:
Should we move this static_assert to the beginning of this function?
I think that would be much clearer to say this function is not for `T==Any`.